standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/client/TestPartitionStat.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.hadoop.hive.metastore.api.ColumnStatisticsObj;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.metastore.api.LongColumnStatsData;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.SetPartitionsStatsRequest;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.client.builder.DatabaseBuilder;
+import org.apache.hadoop.hive.metastore.client.builder.PartitionBuilder;
+import org.apache.hadoop.hive.metastore.client.builder.TableBuilder;
+import org.apache.hadoop.hive.metastore.columnstats.cache.LongColumnStatsDataInspector;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.minihms.AbstractMetaStoreService;
+import org.apache.hadoop.hive.metastore.utils.FileUtils;
+
+import org.apache.thrift.TException;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Tests for updating partition column stats.
+ */
+@RunWith(Parameterized.class)
+@Category(MetastoreCheckinTest.class)
+public class TestPartitionStat extends MetaStoreClientTest {
+  private AbstractMetaStoreService metaStore;
+  private IMetaStoreClient client;
+  private String directSql = "false";
+
+  protected static final String DB_NAME = "test_part_stat";
+  protected static final String TABLE_NAME = "test_part_stat_table";
+  private static final String DEFAULT_COL_TYPE = "int";
+  private static final String PART_COL_NAME = "year";
+  protected static final short MAX = -1;
+  private static final Partition[] PARTITIONS = new Partition[5];
+  public static final String HIVE_ENGINE = "hive";
+
+  @BeforeClass
+  public static void startMetaStores() {
+    Map<MetastoreConf.ConfVars, String> msConf = new HashMap<MetastoreConf.ConfVars, String>();
+    // Enable trash, so it can be tested
+    Map<String, String> extraConf = new HashMap<>();
+    extraConf.put("fs.trash.checkpoint.interval", "30");  // FS_TRASH_CHECKPOINT_INTERVAL_KEY
+    extraConf.put("fs.trash.interval", "30");             // FS_TRASH_INTERVAL_KEY (hadoop-2)
+    startMetaStores(msConf, extraConf);
+  }
+
+  public TestPartitionStat(String name, AbstractMetaStoreService metaStore) {
+    this.metaStore = metaStore;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    // Get new client, store the original value of directSql to restore it back after test.
+    directSql = metaStore.getConf().get(MetastoreConf.ConfVars.TRY_DIRECT_SQL.getVarname());
```

Comment:
I am not sure how `Parameterized` test will behave if we override the `MetaStoreClientTest.getMetaStoreToTest` method, but we might spend a few minutes to try out if we can overrider it with something like this:
```
  @Parameterized.Parameters(name = "{0}")
  public static List<Object[]> getMetaStoreToTest() throws Exception {
    List<Object[]> result = MetaStoreFactoryForTests.getRemoteMetastoresForDirectSQL();
    metaStoreServices = result.stream()
        .map(test -> (AbstractMetaStoreService)test[1])
        .collect(Collectors.toList());
    return result;
  }
```

And implement `MetaStoreFactoryForTests.getRemoteMetastoresForDirectSQL()` so it could return 2 RemoteMetastores. One with directSql on, and one with directSql off.

It might be worth the effort as the resulting code would be much cleaner. We do not have to check for Remote/Embdded metastore and we do not have to set the directSql value manually on/off for the tests

Reply:
// Hack to initialize the stats tables. Not sure why the first test run using remote metastore is failing with
    // error (Unable to update Column stats for  table due to: Table/View 'PART_COL_STATS' does not exist.). This hack
    // has fixed it. All the tests will be executed first using embedded metastore and then with
    // remote metastore. For embedded metastore, TRY_DIRECT_SQL is set to false. For remote metastore TRY_DIRECT_SQL is
    // set to true.
    client.getPartitionColumnStatistics(DB_NAME, TABLE_NAME,
            Collections.singletonList("year=2017"), Collections.singletonList(PART_COL_NAME), HIVE_ENGINE);